elf_load.c

```diff
@@ -311,13 +333,232 @@ static int elf_load_imgcpy(struct elf_prog *elf_prog, Elf *elf,
+		vastart = (uintptr_t)mmap((void *)vastart_aligned,
+					  phdr->p_filesz,
+					  PROT_EXEC | PROT_READ | PROT_WRITE,
+					  MAP_PRIVATE | MAP_FIXED,
+					  fd, phdr->p_offset);
+		if (unlikely(vastart == (uintptr_t)MAP_FAILED)) {
+			uk_pr_err("Failed to mmap the phdr at offset %lu\n",
+				  phdr->p_offset);
+			return (int)vastart;
+		}
+
+		rc = munmap((void *)vastart_old, vastart - vastart_old);
+		if (unlikely(rc)) {
+			uk_pr_err("Failed to unmap [vastart_old, vastart)\n");
+			return rc;
+		}
+	}
+
+	uk_pr_debug("%s: Memory mapped 0x%"PRIx64" - 0x%"PRIx64" to 0x%"PRIx64" - 0x%"PRIx64"\n",
+		    elf_prog->name,
+		    (uint64_t)phdr->p_offset,
+		    (uint64_t)phdr->p_offset + phdr->p_filesz,
+		    (uint64_t)vastart,
+		    (uint64_t)vastart + (uint64_t)phdr->p_filesz);
+
+	vaend_old = PAGE_ALIGN_UP(vastart_old + mmap_len);
+	vaend_new = PAGE_ALIGN_UP(vastart + phdr->p_filesz);
+	if (vaend_old - vaend_new) {
+		rc = munmap((void *)vaend_new, vaend_old - vaend_new);
+		if (unlikely(rc)) {
+			uk_pr_err("Failed to unmap [vaend_new, vaend_old)\n");
+			return rc;
+		}
+	}
+
+	elf_prog->start = vastart;
+	elf_prog->vabase = (void *)vastart;
+	/* We got ehdr.e_entry added initially at the start of elf_load_fd() */
+	elf_prog->entry += (uintptr_t)elf_prog->vabase;
+
+	uk_pr_debug("%s: Program/Library memory region: 0x%"PRIx64"-0x%"PRIx64"\n",
+		    elf_prog->name,
+		    (uint64_t)elf_prog->vabase,
+		    (uint64_t)elf_prog->vabase + elf_prog->valen);
+
+	/* mmap anonymously what we are left if memsz > filesz */
+	vastart = vastart + phdr->p_filesz;
+	vaend = PAGE_ALIGN_UP(vastart + (phdr->p_memsz - phdr->p_filesz));
+	if (vaend > vastart) {
+		rc = elf_load_mmap_filesz_memsz_diff(elf_prog, phdr,
+						     vastart, vaend);
+		if (unlikely(rc)) {
+			uk_pr_err("Failed to map difference between filesz and "
+				  "memsz\n");
+			return rc;
+		}
+	}
+
+	return 0;
+}
+
+/* Use this to mmap every PT_LOAD but the first one */
+static int do_elf_load_fdphdr_not0(struct elf_prog *elf_prog,
+				   GElf_Phdr *phdr, int fd)
+{
+	uintptr_t vastart, vaend;
+	uint64_t delta_p_offset;
+	void *addr;
+	int rc;
+
+	/* If this is not the first PT_LOAD then vabase/start must be != 0 */
+	UK_ASSERT(elf_prog->vabase && elf_prog->start && phdr);
+
+	/* PT_LOAD's paddr may be misaligned, so keep in mind the
+	 * offset from what should have been the aligned paddr to
+	 * what we ended up with.
+	 * As per ELF spec, the file offset and virtual/physical address
+	 * are congruent modulo alignment, which must be multiple of page size.
+	 * Therefore, taking care of the address misalignment will also take
+	 * care of the file misalignment.
+	 */
+	delta_p_offset = phdr->p_paddr - PAGE_ALIGN_DOWN(phdr->p_paddr);
+
+	addr = (void *)PAGE_ALIGN_DOWN((phdr->p_paddr +
+				       (uintptr_t)elf_prog->vabase));
+
+	uk_pr_debug("%s: Memory mapping 0x%"PRIx64" - 0x%"PRIx64" to 0x%"PRIx64" - 0x%"PRIx64"\n",
+		    elf_prog->name,
+		    (uint64_t)phdr->p_offset,
```

Comment:
This should be `phdr->p_offset - delta_p_offset`, and the next one should be `phdr->p_offset + phdr->p_filesz`.